Public/Src/Sandbox/Windows/DetoursServices/PolicyResult.cpp

```diff
@@ -102,3 +103,51 @@ void PolicyResult::ReportIndeterminatePolicyAndSetLastError(FileOperationContext
         ERROR_SUCCESS,
         -1);
 }
+
+bool PolicyResult::AllowWrite() const {
+
+    bool isWriteAllowedByPolicy = (m_policy & FileAccessPolicy_AllowWrite) != 0;
+
+    // Send a special message to managed code if the policy to override allowed writes based on file existence is set
+    // and the write is allowed by policy (for the latter, if the write is denied, there is nothing to override)
+    if (OverrideAllowWriteForExistingFiles() && isWriteAllowedByPolicy) {
+        
+        // Let's check if this path was already checked for allow writes in this process. Observe this structure lifespan is the same 
+        // as the current process so other child processes won't share it. 
+        // But for the current process it will avoid probing the file system over and over for the same path.
+        FilesCheckedForAccess* filesCheckedForWriteAccess = GetGlobalFilesCheckedForAccesses();
+        
+        if (filesCheckedForWriteAccess->TryRegisterPath(m_canonicalizedPath)) {
+            DWORD error = GetLastError();
+
+            // Our ultimate goal is to understand if the path represents a file that was there before the pip started (and therefore blocked for writes).
+            // The existence of the file on disk before the first time the file is written will tell us that. But the problem is that knowing when is the first
+            // time is not trivial: it involves sharing information across child processes.
+            // So what we do is just to emit a special report line with the information of whether the access should be allowed or not, based on existence, from
+            // the perspective of the running process. These special report lines are then processed outside of detours to determine the real first write attempt
+            // Observe this implies that in this case we never block accesses on detours based on file existence, but generate a DFA on managed code
+            bool fileExists = ExistsAsFile(m_canonicalizedPath.GetPathString());
+
+            AccessCheckResult accessCheck = AccessCheckResult(RequestedAccess::Read, ResultAction::Allow, ReportLevel::Ignore);
+            FileOperationContext operationContext(
+                L"FirstAllowWriteCheckInProcess",
+                GENERIC_WRITE,
+                FILE_SHARE_WRITE,
+                OPEN_ALWAYS,
+                FILE_ATTRIBUTE_NORMAL,
+                this->GetCanonicalizedPath().GetPathString());

```

Comment:
I think this should be
```
FileOperationContext::CreateForRead(L""FirstAllowWriteCheckInProcess", this->GetCanonicalizedPath().GetPathString());
```

Reply:
Cool, didn't know we had that. Will change.

---
In reply to: [299200131](https://github.com/microsoft/BuildXL/pull/536#discussion_r299200131) [](ancestors = 299200131)